walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,95 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.lang.Math;
+import java.io.PrintStream;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-  }
+    public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+        if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+            return Arrays.asList();
+        }
+        if (events.size() == 0) {
+            return Arrays.asList(TimeRange.WHOLE_DAY);
+        }
+        List<String> allAttendees = new ArrayList<>();
+        allAttendees.addAll(request.getAttendees());
+        allAttendees.addAll(request.getOptionalAttendees());
+
+        List<TimeRange> timeRangesIngoreOptional = new ArrayList<>();
+        List<TimeRange> timeRangesIncludeOptional = new ArrayList<>();
+        for (Event item : events) {
+            for (String s : request.getAttendees()) {
+                if (item.getAttendees().contains(s)) {
+                    timeRangesIngoreOptional.add(item.getWhen());
+                    break;
+                }
+            }
+            for (String s : allAttendees) {
+                if (item.getAttendees().contains(s)) {
+                    timeRangesIncludeOptional.add(item.getWhen());
+                    break;
+                }
+            }
+        }
+
+        Collection<TimeRange> ignoreOptional = availableTimeRange(timeRangesIngoreOptional, request);
+        Collection<TimeRange> includeOptional = availableTimeRange(timeRangesIncludeOptional, request);
+        if (request.getAttendees().size() == 0) {
+            return includeOptional;
+        }
+        if (includeOptional.size() != 0) {
+            return includeOptional;
+        }
+        return ignoreOptional;
+    }
+
+    // return timeRanges that satisfy meeting duration from all available timeRanges
+    public Collection<TimeRange> availableTimeRange(List<TimeRange> timeRanges, MeetingRequest request) {
+        Collections.sort(timeRanges, TimeRange.ORDER_BY_START);
+        int len = timeRanges.size();
+        if (len == 0) {
+            return Arrays.asList(TimeRange.WHOLE_DAY);
+        }
+        if (len == 1) {
+            return Arrays.asList(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, timeRanges.get(0).start(), false),
+                TimeRange.fromStartEnd(timeRanges.get(0).end(), TimeRange.END_OF_DAY, true));
+        }
```

Comment:
I feel like you shouldn't need this, the len == 1 case should be handled automatically already